man/xmlclient.8
.\" Copyright (C) 2010 Internet Systems Consortium, Inc. ("ISC")
.\" 
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS" AND ISC DISCLAIMS ALL WARRANTIES WITH
.\" REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS. IN NO EVENT SHALL ISC BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
.\" LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
.\" OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
.\" PERFORMANCE OF THIS SOFTWARE.
.\"
.\" $Id$
.\"
.hy 0
.ad l
'\" t
.\"     Title: xmlclient.py
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: October 10, 2010
.\"    Manual: AFTR
.\"    Source: AFTR
.\"  Language: English
.\"
.TH "XMLCLIENT\&.PY" "8" "October 10, 2010" "AFTR" "AFTR"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
xmlclient.py \- remote configuration client for aftr
.SH "SYNOPSIS"
.HP 13
\fBxmlclient\&.py\fR \fIaftr\-addr\fR [\fIcommand\fR]
.SH "OPTIONS"
.PP
aftr\-addr
.RS 4
This is the address (IPv4 or IPv6) the the target AFTR\&.
.RE
.SH "COMMANDS"
.PP
create \fIuser\-ipv6\fR \fIprotocol\fR \fIsrc\-ipv4\fR \fIsrc\-port\fR \fInat\-ipv4\fR \fInat\-port\fR
.RS 4
This requests the aftr to create a port mapping\&.
.RE
.PP
create \fIuser\-ipv6\fR \fInat\-ipv4\fR
.RS 4
This requests the aftr to create a tunnel entry, using
\fInat\-ipv4\fR
as the natted IPv4 address for all future port mappings on this tunnel (dynamic as well as static)\&.
.RE
.PP
delete \fIuser\-ipv6\fR \fIprotocol\fR \fIsrc\-ipv4\fR \fIsrc\-port\fR \fInat\-ipv4\fR \fInat\-port\fR
.RS 4
.RE
.PP
delete \fIuser\-ipv6\fR \fIprotocol\fR \fIsrc\-ipv4\fR \fIsrc\-port\fR
.RS 4
.RE
.PP
delete \fIprotocol\fR \fInat\-ipv4\fR \fInat\-port\fR
.RS 4
These three forms of the
\fBdelete\fR
command all request the aftr to delete a port mapping\&. The mapping can be fully specified (first form), but a mapping can also be uniquely identifed by either internal parameters (second form) or external parameters (third form)\&.
.RE
.PP
delete \fIuser\-ipv6\fR
.RS 4
This requests the aftr to delete all port mappings (dynamic as well as static) and other state associated with the given tunnel address\&. This is often done prior to moving the customer to a new natted IPv4 address\&.
.RE
.PP
flush
.RS 4
This requests the aftr to remove all static port mappings and configured tunnel entries\&. Note that this is a very drastic action, and should only be undertaken if (for example) the aftr configuration is seriously out of sync with the provisioning system\&.
.RE
.PP
get \fIuser\-ipv6\fR
.RS 4
This requests the aftr to report all static port mappings associated with the given tunnel address\&.
.RE
.PP
get
.RS 4
This requests the aftr to report all static port mappings, and all configured tunnels without static port mappings\&.
.RE
.SH "SCRIPTING"
.PP
If no commands are given on the command line,
\fBxmlclient\&.py\fR
will read commands from stdin\&. This allows the provisioning system to accumulate changes for a given AFTR, and send them all at once\&.
.PP
In general, it is probably easier for the provisioning system to send requests immediately, and get replies immediately\&. However, some operators may prefer to batch up requests, and this method sends multiple requests over an open connection, without having to establish a connection for each request\&.
.PP
Example:
.br
\fBxmlclient\&.py 2001::500 <script\fR
.PP
where
script
contains:
.br
\fBcreate 2001::525a:8c5a:30d4:e36e tcp 192\&.168\&.0\&.88 6265 198\&.18\&.200\&.174 5005\fR
.br
\fBcreate 2001::835c:1eff:8d66:22fc tcp 192\&.168\&.1\&.138 3877 198\&.18\&.200\&.121 5572\fR
.br
\fBcreate 2001::e3:9a2f:8abf:40de:2d87 udp 192\&.168\&.0\&.92 7356 198\&.18\&.200\&.149 5547\fR
.br
.SH "SEE ALSO"
.PP
\fBxmlconf\fR(8)
.SH "AUTHOR"
.PP
Internet Systems Consortium
.SH "COPYRIGHT"
.br
Copyright \(co 2010 Internet Systems Consortium, Inc. ("ISC")
.br


CHANGES
	--- aftr-1.1rc1 released ---

 119.	[bug]		Update/fix xpmpd.py. [RT #22370]

	--- aftr-1.1b2 released ---

 118.	[doc]		Updated documentation, added book.txt,
			streamlined example config files.

 117.	[func]		Major improvements to the XML tools.

 116.	[bug]		Protect nonat entries against tunnel operations,
			and add nonat ICMP translation. [RT #22330]

 115.	[bug]		Set the correct next-header value after IPv6
			reassembly. [RT #22238]

 114.	[bug]		Use UDP=1/TCP=2 in UPnP IWF. [RT #22100]

 113.	[func]		Add 'reload succeeded' message. [RT #21596]

 112.	[bug]		Protect the range loop against clock jumping
			backward.

	--- aftr-1.1b1 released ---

 111.	[func]		Add XML-based remote configuration tools. [RT #20379]

 110.	[func]		Don't list stale entries in 'list nat global'.
			[RT #21630]

 109.	[bug]		Per session command buffer. [RT #21599]

 108.	[func]		Extend the try tunnel command. [RT #21566]

 107.	[func]		Merge UPNP IGD proxy. [RT #21184]

 106.	[func]		Add a search in logs/traces tool. [RT #21185]

	--- aftr-1.0.1 released ---

 105.   [func]		Silence spurious gcc warnings.

 104.	[doc]		Slightly improve the documentation. [RT #20906]

	--- aftr-1.0 released ---

 103.	[doc]		Give configuration examples for an AFTR which
			shares one public address for everything
			(confs/*shareone* files). [RT #20770]

 102.	[func]		Add acl4/private default. [RT #20726]

 101.	[bug]		Check the IPv4 source address of packets from a
			no-nat B4. [RT #20684]

 100.	[func]		Shorter (1ms vs 1s)  main loop select() timeout
			when some housekeeping has to be done. [RT #20684]

  99.	[doc]		Document ACL processing. [RT #20684]

  98.	[doc]		Add a README in confs listing the files. [RT #20684]

  97.	[func]		Add 'default private' command and improve acl6
			command. [RT #20684]

  96.	[doc]		Documentation reorganization: source in doc,
			manuals in man, HTML in html. [RT #20684]

	--- aftr-1.0b1 released ---

  95.	[func]		Add support for extended NAT-PMP. [RT #20377]

  94.	[func]		Add an echo command. [RT #20653]

  93.	[bug]		Protect against unexpected 'session close'.
			[RT #20636]

  92.	[func]		Add a try command and notification of tunnel removal
			for NAT-PMP/UPnP IGD daemon benefit. [RT #20631]

  91.	[doc]		Create reference and manuals in docbook. [RT #20483]

  90.	[func]		Rename cgn* to aftr*. [RT #20348]

  89.	[func]		Promote a lot of compile time options to
			section zero defaults. [RT #19993]

  88.	[func]		Add check, fork and abort commands. Add
			compile time SIGNSHDR option for a
			signature header in structures.
			Fix a lot of bugs... [RT #20537]

  87.	[func]		Implement the reload command. [RT #20275]

  86.	[func]		Add an atomic load file command (it is a
			form of input redirection). [RT #20539]

  85.	[func]		Add debug/list/... help sub-commands. [RT #20543]

  84.	[func]		Add control channel (aka. session) context
			(with some commands renaming). [RT #20481]

  83.	[func]		Add some command line options. [RT #20536]

  82.	[doc]		Add a section about syslog in README. [RT #20528]

  81.	[bug]		Eats extra spaces in command/config using
			strtok_r(). [RT #20529]

  80.	[func]		Add a PF_UNIX control channel. [RT #20530]

  79.	[bug]		Add srcidx in list tunnel. [RT #20471]

  78.	[func]		Add a check for dynamic/static conflicts
			(i.e., a static NAT entry configured in
			a dynamic port range). [RT #20468]

  77.	[func]		Add -c for alternate config. [RT #20376]

  76.	[func]		Merge start/stop scripts. [RT #20427]

  75.	[func]		Add control channels. [RT #20276]

  74.	[func]		Split config commands into sections. [RT #20271]

  73.	[func]		Deal with up to QUANTUM packets in main loop
			before housekeeping. [RT #20373]

  72.	[func]		Recode list operations with macros. [RT #20347]

  71.	[func]		Check for counters which go under zero. [RT #20349]

  70.	[func]		Add the reboot command. [RT #20273]

  69.	[func]		Use syslog for logxxx(), including tracing
			(so the trace command was removed). [RT #20272]

  68.	[func]		Rename address and port commands. [RT #19994]

  67.	[func]		Switch to pure allocation buckets. [RT #20345]

  66.	[bug]		Release bucket when new_nat() fails. [RT #20199]

  65.	[func]		Add BUCKET_DEBUG compile flag which enables
			"debug bucket all".

  64.	[func]		Make LIFETIMEs compile time tunables.

  63.	[func]		Add TRACE_NAT compile flag (default is undef).

  62.	[bug]		Prompt not-yet empty bucket.

  61.	[func]		Add an option IPv4 source address in tunnel command
			[RT #19992].

  60.	[func]		Add NAT creation rate limit (per tunnel/protocol).
			Not yet tested.

  59.	[func]		TCP ALG is finished.

        --- cgn-usermode-snapshot-20090709 released ---

  58.	[doc]		Warn about the "deny all" of IPv6 ACL.

  57.	[bug]		Another fix in splay tree code for wilcards.
			FTP ALG works if there is no retransmissions.

  56.	[bug]		Begin to fix the FTP ALG (not finished).

  55.	[bug]		Fix a typo in the splay tree integration.
			src1<->cgn tests run again.

  54.	[bug]		The 192.0.0.0 prefix is a /29.

  53.	[func]		First version of FTP ALG.

  52.	[func]		Implement No-NAT.

  51.	[func]		Add TCP MSS patching for PRR (aka A+P).

  50.	[bug]		Fix two bucket bugs.

  49.	[func]		Add IPv6 ACL.

  48.	[bug]		Fix crash on some truncated commands.

  47.	[func]		Add "debug bucket <addr>" command.

  46.	[bug]		Fix ICMP echo translation.

  45.	[func]		Add port ranges and buckets.

  44.	[func]		Add show [dropped|stat] aliases.

  43.	[func]		Add dropped counters (and "debug dropped" command).

  42.	[func]		Add NAT entry counters & co.

  41.	[func]		Add NAT entry creation/deletion counters and rates.

  40.	[func]		Add per tunnel max fragment list length and
			defrag*() counters.

  39.	[func]		Remove port affinity.

  38.	[func]		Make MAXNATCNT visible.

  37.	[func]		Add rates (with 1/5/15 mn decays).

  36.	[func]		Add 192.0.0.0/24 from the unpublished I-D.

  35.	[func]		Implement per-tunnel debug counters.


xml/xmlclient.py
#!/usr/bin/env python

# Copyright (C) 2010  Internet Systems Consortium, Inc. ("ISC")
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND ISC DISCLAIMS ALL WARRANTIES WITH
# REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
# AND FITNESS.  IN NO EVENT SHALL ISC BE LIABLE FOR ANY SPECIAL, DIRECT,
# INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
# LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
# OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
# PERFORMANCE OF THIS SOFTWARE.
#

"""
Client side of AFTR remote configuration

aftrconf-client <aftr-addr> <command>

commands:
create <user-ipv6> <protocol> <src-ipv4> <src-port> <nat-ipv4> <nat-port>
create <user-ipv6> <nat-ipv4>
delete <user-ipv6> <protocol> <src-ipv4> <src-port> <nat-ipv4> <nat-port>
delete <user-ipv6> <protocol> <src-ipv4> <src-port>
delete <protocol> <nat-ipv4> <nat-port>
delete <user-ipv6>
flush
get <user-ipv6>
get
"""

__version__ = '$Id: xmlclient.py 945 2010-10-19 20:53:12Z pselkirk $'

import sys
import socket
from lxml import etree

TRANSPORT = 'http'
#TRANSPORT = 'socket'
#TRANSPORT = 'debug'

NSMAP = {'aftr': 'http://aftr.isc.org/mapping/1.0'}

SOCKPORT = 4146
HTTPPORT = 4148

VERBOSE = False

def debug(msg):
    """debug helper"""
    if VERBOSE:
        print msg

if TRANSPORT == 'socket':
    SOCK = None

    def send(addr, request):
        """ send <rpc> request """
        global SOCK
        if SOCK is None:
            addrlist = socket.getaddrinfo(addr, SOCKPORT)
            try:
                SOCK = socket.socket(addrlist[0][0],
                                     socket.SOCK_STREAM,
                                     socket.IPPROTO_TCP)
            except socket.error as err:
                print 'ERROR:', err[1]
                exit()
            try:
                SOCK.connect(addrlist[0][4])
            except socket.error as err:
                print 'ERROR:', err[1]
                exit()
        try:
            SOCK.sendall(etree.tostring(request,
                                     encoding='UTF-8',
                                     xml_declaration=True,
                                     pretty_print=True))
        except socket.error as err:
            print 'ERROR:', err[1]
            exit()
        response = ''
        while True:
            try:
                buf = SOCK.recv(1024)
                if len(buf) == 0:
                    break
                response += buf
            except socket.error:
                # host unreachable, connection refused, etc
                pass
        processresponse(response)

elif TRANSPORT == 'http':
    import httplib
    CONN = None

    def send(addr, request):
        """ send <rpc> request """
        global CONN
        if CONN is None:
            addrlist = socket.getaddrinfo(addr, HTTPPORT)
            try:
                CONN = httplib.HTTPConnection(addrlist[0][4][0],
                                              addrlist[0][4][1])
            except socket.error as err:
                print 'ERROR:', err[1]
                exit()
        try:
            CONN.request('POST', '',
                         etree.tostring(request, pretty_print=True))
        except socket.error as err:
            print 'ERROR:', err[1]
            exit()
        response = CONN.getresponse()
        debug('%s %s' % (response.status, response.reason))
        buf = response.read()
        processresponse(buf)

else:	# TRANSPORT == 'debug'
    def send(addr, request):
        """ send <rpc> request """
        print 'addr:', addr
        # debugging: just print the request
        # (this can be piped to the server)
        print etree.tostring(request,
                             encoding='UTF-8',
                             xml_declaration=True,
                             pretty_print=True)

def processresponse(response):
    """process the response"""
    parser = etree.XMLParser(remove_blank_text=True,
                             remove_comments=True)
    try:
        #debug(response)
        root = etree.fromstring(response, parser)
    except etree.XMLSyntaxError:
        # XXX syslog?
        print 'error parsing response'
        raise
    if root.tag != 'rpc-reply':
        print 'invalid document from aftr:'
        print etree.tostring(root,
                             encoding='UTF-8',
                             xml_declaration=True,
                             pretty_print=True)
    else:
        for entry in root:
            if entry.tag == 'ok':
                print 'ok'
            elif entry.tag == 'rpc-error':
                print entry.text
            elif entry.tag == 'conf':
                for elem in entry:
                    if elem.tag == 'tunnelEntry':
                        tunnel = elem.find('tunnel')
                        addr = elem.find('nattedAddress')
                        if (tunnel is None) or (addr is None):
                            debug('missing element')
                            continue
                        print 'tunnel', tunnel.text, addr.text
                    elif elem.tag == 'natEntry':
                        tunnel = elem.find('tunnel')
                        proto = elem.find('protocol')
                        saddr = elem.find('sourceAddress')
                        sport = elem.find('sourcePort')
                        naddr = elem.find('nattedAddress')
                        nport = elem.find('nattedPort')
                        if (tunnel is None) or (proto is None) or \
                                (saddr is None) or (sport is None) or \
                                (naddr is None) or (nport is None):
                            debug('missing element')
                            continue
                        print 'nat', tunnel.text, proto.text, \
                            saddr.text, sport.text, naddr.text, nport.text
                    else:
                        print 'unexpected <conf> content: ', elem.tag
                        continue
            else:
                print 'unexpected rpc-reply:'
                print etree.tostring(entry,
                                     encoding='UTF-8',
                                     xml_declaration=True,
                                     pretty_print=True)

def parsecreate(args, request):
    """ parse a create command """
    # multiple forms of create:
    if len(args) == 6:
        # XXX sanity check args?
        operation = etree.SubElement(request, 'create')
        etree.SubElement(operation, 'tunnel').text = args[0]
        etree.SubElement(operation, 'protocol').text = args[1]
        etree.SubElement(operation, 'sourceAddress').text = args[2]
        etree.SubElement(operation, 'sourcePort').text = args[3]
        etree.SubElement(operation, 'nattedAddress').text = args[4]
        etree.SubElement(operation, 'nattedPort').text = args[5]
    elif len(args) == 2:
        # XXX sanity check args?
        operation = etree.SubElement(request, 'create')
        etree.SubElement(operation, 'tunnel').text = args[0]
        etree.SubElement(operation, 'nattedAddress').text = args[1]
    else:
        print 'usage: create <user-ipv6> <protocol>', \
            '<src-ipv4> <src-port> <nat-ipv4> <nat-port>\n', \
            '       create <user-ipv6> <nat-ipv4>'
        return False
    return True

def parsedelete(args, request):
    """ parse a delete command """
    # multiple forms of delete:
    if len(args) == 6:
        # XXX sanity check args?
        operation = etree.SubElement(request, 'delete')
        etree.SubElement(operation, 'tunnel').text = args[0]
        etree.SubElement(operation, 'protocol').text = args[1]
        etree.SubElement(operation, 'sourceAddress').text = args[2]
        etree.SubElement(operation, 'sourcePort').text = args[3]
        etree.SubElement(operation, 'nattedAddress').text = args[4]
        etree.SubElement(operation, 'nattedPort').text = args[5]
    elif len(args) == 4:
        # XXX sanity check args?
        operation = etree.SubElement(request, 'delete')
        etree.SubElement(operation, 'tunnel').text = args[0]
        etree.SubElement(operation, 'protocol').text = args[1]
        etree.SubElement(operation, 'sourceAddress').text = args[2]
        etree.SubElement(operation, 'sourcePort').text = args[3]
    elif len(args) == 3:
        # XXX sanity check args?
        operation = etree.SubElement(request, 'delete')
        etree.SubElement(operation, 'protocol').text = args[0]
        etree.SubElement(operation, 'nattedAddress').text = args[1]
        etree.SubElement(operation, 'nattedPort').text = args[2]
    elif len(args) == 1:
        # XXX sanity check args?
        operation = etree.SubElement(request, 'delete')
        etree.SubElement(operation, 'tunnel').text = args[0]
    else:
        print 'usage: delete <user-ipv6> <protocol> <src-ipv4> ', \
            '<src-port> <nat-ipv4> <nat-port>\n', \
            '       delete <user-ipv6> <protocol> <src-ipv4> <src-port>\n', \
            '       delete <protocol> <nat-ipv4> <nat-port>\n', \
            '       delete <user-ipv6>'
        return False
    return True

def parseflush(args, request):
    """ parse a flush command """
    if len(args) != 0:
        print 'usage: flush'
        return False
    etree.SubElement(request, 'flush')
    return True

def parseget(args, request):
    """ parse a get command """
    # two forms of get:
    if len(args) == 1:
        # XXX sanity check args?
        operation = etree.SubElement(request, 'get')
        operation.set('tunnel', args[0])
    elif len(args) == 0:
        etree.SubElement(request, 'get')
    else:
        print 'usage: get [tunnel]'
        return False
    return True

def dispatch(args):
    """parse and send one command to the aftr"""
    request = etree.Element('rpc', nsmap=NSMAP)
    if args[2] == 'create':
        ret = parsecreate(args[3:], request)
    elif args[2] == 'delete':
        ret = parsedelete(args[3:], request)
    elif args[2] == 'get':
        ret = parseget(args[3:], request)
    elif args[2] == 'flush':
        ret = parseflush(args[3:], request)
    else:
        print 'unknown command "%s"' % args[2]
        ret = False
    if ret:
        send(args[1], request)

def main(args):
    """main"""
    if len(args) < 2:
        print 'usage:', args[0], '<aftr-addr> get|create|delete|flush'
        return
    if len(args) == 2:
        # read commands from a script file
        for line in sys.stdin:
            print '[' + line.strip() + ']'
            largs = line.split()
            # skip blank lines
            if len(largs) < 1:
                continue
            dispatch(args + largs)
    else:
        # single command
        dispatch(args)

main(sys.argv)
